R/quosure-join.R

```diff
@@ -0,0 +1,115 @@
+#' Join two `Quosure` objects
+#'
+#' Combine two `Quosure` object by merging their environments and the code.
+#' Not every `Quosure` objects can be combined:
+#' - if their environments contains objects of the same name but not identical
+#' - if `y` has unique code element placed before common element. This means that `y`
+#' in the environment of the `y` was evaluated some extra code before which can influence
+#' reproducibility
+#' - more cases to be done
+#' @param x (`Quosure`)
+#' @param y (`Quosure`)
+#' @examples
+#' q1 <- new_quosure(
+#'   code = c(iris1 = "iris1 <- iris", mtcars1 = "mtcars1 <- mtcars"),
+#'   env = list2env(list(
+#'     iris1 = iris,
+#'     mtcars1 = mtcars
+#'   ))
+#' )
+#' q2 <- q1
+#' q1 <- eval_code(q1, "iris2 <- iris")
+#' q2 <- eval_code(q2, "mtcars2 <- mtcars")
+#' qq <- join(q1, q2)
+#' get_code(qq)
+#' @export
+setGeneric("join", function(x, y) {
+  standardGeneric("join")
+})
+
+#' @rdname join
+#' @export
+setMethod("join", signature = c("Quosure", "Quosure"), function(x, y) {
+  join_validation <- check_joinable(x, y)
+
+  # join expressions
+  if (!isTRUE(join_validation)) {
+    stop(join_validation)
+  }
+
+  id_unique <- !y@id %in% x@id
+  x@id <- c(x@id, y@id[id_unique])
+  x@code <- .keep_code_name_unique(x@code, y@code[id_unique])
+
+  # insert (and overwrite) objects from env2 to env
+  x@env <- .copy_env(x@env)
+  lapply(ls(y@env, all.names = TRUE), function(xx) assign(xx, get(xx, y@env), x@env))
+
+  x
+})
+
+#' @rdname join
+#' @export
+setMethod("join", signature = "quosure.error", function(x, y) {
+  x
+})
+
+#' @rdname join
+#' @export
+setMethod("join", signature = c("Quosure", "quosure.error"), function(x, y) {
+  y
+})
+
+#' If two `Quosure` can be joined
+#'
+#' Checks if two `Quosure` objects can be combined.
+#' They can't be combined if (and):
+#' - both share the same code (identified by `id`)
+#' - indices of the shared code are not consecutive or don't start from 1
+#' @param x (`Quosure`)
+#' @param y (`Quosure`)
+#' @return `TRUE` if able to join or `character` used to print error message.
```

Comment:
it's always good to assure that function always returns the same data type. This would prevent you doing `expect_true(foo(...))` and `expect_match(foo(...), "...")` at the same time.

This is not the first time where we are facing the problem of returning a character message if something else TRUE. I suggest to always return boolean with additional attribute. Alternatively you might want to use [`checkmate` assertion collections](https://mllg.github.io/checkmate/reference/AssertCollection.html).

Reply:
I agree. As a user it can be annoying when a function returns 2 types of objects. ('user' includes developer w/in package)

Reply:
This function is internal, user shouldn't use this function at all. `.check_joinable` is used only in the `join` in following way:
```r
  join_validation <- .check_joinable(x, y)

  # join expressions
  if (!isTRUE(join_validation)) {
    stop(join_validation)
  }
```

Alternatively, we can move `stop` to the .check_joinable and forget the whole thing